app/Http/Controllers/Admin/CategoryController.php

```diff
@@ -0,0 +1,158 @@
+            $slug = Str::slug($request->title);
+            $category = array_merge($category, ['slug' => $slug]);
+        }
+
+        $category = Category::create($category);
+        if (!empty($request->file('image'))) {
+            $category->addMediaFromRequest('image')
+                ->preservingOriginal()
+                ->toMediaCollection('categories');
+        }
+        return redirect()->route('category.index')->with('message', 'Категория успешно добавлена');
+    }
+
+    /**
+     * Display the specified resource.
+     *
+     * @param int $id
+     * @return Response
+     */
+    public function show($id)
+    {
+        //
+    }
+
+    /**
+     * Show the form for editing the specified resource.
+     *
+     * @param Category $category
+     * @return Response
+     */
+    public function edit(Category $category)
+    {
+        $image = $category->getFirstMedia('categories');
+        if (!empty($image)) {
+            $image = $image->getUrl();
+        } else {
+            $image = Storage::url('0/no_photo.png');
+        }
+        $parentCategories = Category::where('company_id', $category->company_id)->whereNull('parent_id')->get(['id', 'title', 'company_id']);
+
+        return view('admin/category/edit', [
+            'category' => $category,
+            'parentCategories' => $parentCategories,
+            'image' => $image]);
+    }
+
+    /**
+     * Update the specified resource in storage.
+     *
+     * @param StoreCategory $request
+     * @param Category $category
+     * @return Response
+     * @throws Exception
+     */
+    public function update(EditCategory $request, Category $category)
+    {
+        $data = $request->validated();
+        if (empty($request->slug)) {
+            $slug = Str::slug($request->title);
+            $data = array_merge($data, ['slug' => $slug]);
+        }
+
+        $category->update($data);
+
+        if (!empty($request->file('image'))) {
+            if (!empty($category->getFirstMedia('categories'))) {
+                $category->getFirstMedia('categories')->delete();
+            }
+            $category->addMediaFromRequest('image')
+                ->preservingOriginal()
+                ->toMediaCollection('categories');
+        }
+        return redirect()->route('category.index')->with('message', 'успешно изменено!!!');
+    }
+
+    /**
+     * Remove the specified resource from storage.
+     *
+     * @param Category $category
+     * @return Response
+     * @throws Exception
+     */
+    public function destroy(Category $category)
+    {
+        if (count($category->nestedCategories) == 0 && count($category->sets) == 0) {
+            $category->delete();
+            return redirect()->route('category.index')->with('message', 'успешно удалено!!!');
+        } else {
+            return redirect()->route('category.index')->with('message', 'Нельзя удалить, пока есть зависимые категории или комплекты');
+        }
+    }
```

Comment:
Смотри, когда ты пишешь $category->nestedCategories, ларавель делает запрос к БД и вытаскивает все вложенные категории, создает для каждой модель и складывает это всё в коллекцию, т.е. это всё расходует память. Тоже самое справедливо для $category->sets. Это ничего страшного и в нашем проекте оно работает нормально. Но в серьезных проектах так делать не надо, потому что нам тут не нужны все эти модели и коллекции в памяти, мы просто хотим посчитать кол-во строк в БД! Для этого можно просто сделать два запроса COUNT и не грузить в память лишнее. Ну и сам код легко можно сделать менее громоздким.

```suggestion
    public function destroy(Category $category)
    {
        if ($category->nestedCategories()->count() === 0 && $category->sets()->count() === 0) {
            $category->delete();
            $msg = 'Категория удалена';
        }
        
           return redirect()->route('category.index')->with('message',  $msg ?? 'Нельзя удалить, пока есть зависимые категории или комплекты');
        }
    }
```